### ISSUE TITLE
Data source depends_on fails in refresh if file modified outside of Terraform

### ISSUE THREAD

**Issue by pselle** [hashicorp/terraform#25463]
### Terraform Version

```
Terraform v0.13.0-dev
```

### Terraform Configuration Files

```terraform
resource "local_file" "apple" {
    content     = "apple"
    filename = "${path.root}/apple.txt"
}

data "local_file" "apple_d" {
    filename = "${path.root}/apple.txt"
    depends_on = [local_file.apple]
}

output "fruit" {
  value = data.local_file.apple_d.content
}
```

### Expected Behavior

Terraform should create the missing file

### Actual Behavior

Terraform errors in the refresh phase
```
local_file.apple: Refreshing state... [id=d0be2dc421be4fcd0172e5afceea3970e2f3d940]
data.local_file.apple_d: Refreshing state... [id=d0be2dc421be4fcd0172e5afceea3970e2f3d940]

Error: open ./apple.txt: no such file or directory

  on main.tf line 6, in data "local_file" "apple_d":
   6: data "local_file" "apple_d" {
```

Note: This was discovered when using module expansion between modules using `depends_on` where data sources are used -- I recommend testing any fix with module `depends_on` as well.

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. `rm apple.txt`
4. `terraform apply`

**Comment by jbardin** [comment]
Unfortunately, this would still be expected for now, until we can conditionally refresh things.

There are often changes that can be done outside of terraform that will break the current state, and one would need to fix it manually either by updating the state to match reality, or by changing the config. In this case it's fixed by removing the record of the data source that no longer exists via `terraform state rm data.local_file.apple_d`.

While we have not previously allowed data sources to fail silently, we may be able to look into making an exception when `depends_on` is set at all. I'm not sure I like the idea of allowing it to fail during refresh, but I think that if the data source is used in a manner that allowed `depends_on` in the first place, it would be safe to delay it again for a later plan.

**Comment by pselle** [comment]
@jbardin I should have included this in the original issue, but this works in 0.12 and is broken in 0.13, so I consider it a bug.

**Comment by apparentlymart** [comment]
I guess what's going on here is that our previous "clumsy" data source `depends_on` behavior would've forced this data source to be read during the apply phase regardless of what is planned for anything else, and thus the data resource would've been processed after the Create action for the resource, but our new handling of `depends_on` only defers if we can see that there's a change pending for the resource.

However, based on the behavior I think we _intended_ here I would have still expected this to work and to have deferred the data resource until the apply phase because:

1. The refresh walk should see that `depends_on` is set for the data resource and thus know it can't safely deal with the read during the refresh walk -- we need to see the plan before we can decide.
2. The plan walk should then see that the data resource depends on the managed resource and so plan the managed resource first. Once it gets to planning the data resource, it'd find a `Create` plan for the managed resource and should thus see that the read must be deferred until the apply phase, and so generate a `Read` plan for the data resource.
3. Then finally, during the apply walk, again the data resource depends on the managed resource so the manage resource `Create` should happen first, followed by the data resource `Read`, thus getting the expected result.

Since we can see in this report that the data resource read is happening during the refresh walk, it seems like step 1 is where the problem is arising here: for some reason it is not detecting that the decision must be deferred until the plan walk. If that step were being honored then I'd expect the rest to unfold correctly, because both the plan walk and the apply walk have enough information to handle the situation correctly.

Once we merge the refresh and plan walks this'll become moot of course, but I've seen many examples in Q&A of folks using `depends_on` for data resources so I think it's important to make sure this is still working at least as clumsily as 0.12 handled it (conservatively deferring a `depends_on` data resource to the apply walk in all cases). _Ideally_ it'd behave like described above with the ability to still deal with a read during the plan phase if there's no pending action for the managed resource, but that's not a headline feature we've advertised for 0.13 so it could wait until we merge refresh+plan if it's not practical to implement that right now.

**Comment by jbardin** [comment]
I don't think there will be a simple path to trading this new broken behavior for the old broken behavior, since the ability to compare the refreshed state to the proposed state is required to be able to plan a data source, in the same manner that we plan a managed resource. In that sense, we would need a way to generate a plan out from refresh, which is essentially what we intend to do with the refresh+plan merge. 

I'm not certain that having a data source which is going to fail entirely when a dependency changes is going to be very common. The majority of instances will simply refresh then refresh again with new config values, since the prior config must have applied successfully to get the existing state. Since this is likely an uncommon case, perhaps allowing a data source to fail silently in Refresh if there are any references in `depends_on` would be an OK concession.

**Comment by apparentlymart** [comment]
If the refresh walk would be able to modify its handling of _errors_ from the data resource read operation based on a non-empty `depends_on`, would it be a similar amount of effort to make it skip the read altogether in that case, or does skipping the read introduce additional complexity that doesn't apply to reading and then silently tolerating a failure?

**Comment by jbardin** [comment]
The problem is that (currently) the only way to communicate something from refresh to plan is through the state, and we do that for data sources by removing the prior state from the plan, and a functioning data source must still be read during refresh, because you can't save the state during plan. Not refreshing when there's a `depends_on` would in effect bring back the old behavior where the data source always produce a diff, but that is the part we can't bring back because it too severely limits the ability to use `depends_on` with modules. 

I vaguely recall trying to use the the internal `ObjectPlanned` status in the state to communicate the need for a plan; I may have dropped it because it didn't seem necessary to change the state format at the time, but it could be an option. The more I think about it however, the more I think simply logging the error during refresh would be a workable solution here, since that doesn't require any further change to the refresh+plan+apply mechanisms.

**Comment by apparentlymart** [comment]
It seems like we have a few different options here, none of which are particularly awesome:

1. Keep the current behavior: it will work for most "normal" situations where a data source is fetching something that exists in a remote system or generates a result entirely from its inputs. However, it will fail in any case where the object being read is also managed by this same configuration and hasn't been created yet, either because this is the first `terraform apply` or because the object was deleted outside of Terraform.

    Although we do advise not managing and reading the same object in the same configuration, my instincts from Q&A lead me to suspect that this is a common approach in existing configurations, perhaps due to former limitations in 0.11 that made it impossible to pass whole objects between modules.

2. Return to the previous behavior where any explicit `depends_on` on a data source causes it to always be deferred to apply, and extend that behavior to now include indirect `depends_on` via the containing module too.

    As @jbardin noted, this would make `depends_on` for modules very annoying to use for any module that contains at least one data resource, because the module-level `depends_on` would behave like a resource-level `depends_on` for the data resource and thus force it to defer to apply time.

    On the other hand, this is the behavior whose consequences we're most familiar with because it's the behavior we've observed under 0.12. We're planning to merge the refresh and plan walks in a future release, which would allow us to get the behavior we were aiming for here: making `depends_on` defer a data resource to apply only if one of the specified explicit dependencies is or transitively depends on a resource which already has a planned action.

3. As a compromise, make a read failure during refresh _not_ fail the refresh but instead force the data resource to be deferred to the apply phase as if it contained unknown values.

    In the main case the data resource would then tend to fail again at apply time, because nothing will have changed, but in the problematic case (configuration is both managing and reading the same object) it would then probably _work_, because the object will exist by the time the read happens at apply time.

    This is tricky because we currently have no way to talk about this situation in the user model, and so we'd either need to do it totally silently or we'd need to do something awkward, like turning the error diagnostic into a warning diagnostic with an extra note like "Because of this error, the data resource will be deferred until the apply phase."

    It's also tricky for data sources that violate the usual contract for data sources by having externally-visible side-effects. For example, the `template_dir` data source could fill a directory with files and then fail partway through, creating a strange situation where the operation it represents has partially completed. For another example, the `vault` provider data sources could end up creating leases that are immediately discarded without using them, and create entirely new leases during the apply phase. Neither of these situations seems super likely: they are both "partial fail" situations where an error occurs after some side-effects have already taken place, whereas it's more likely for them to either entirely fail or entirely succeed.

(I'm intentionally excluding some options that would require original design and refactoring, because those solutions would not be appropriate at the current stage in the release cycle. For example, making the plan phase be able to take additional "refresh-like" actions to contribute to the prior state could work in theory, but that starts to chew off a chunk of the "merge refresh and plan" project and so is, I think, too risky to jump into now.)

Are there other options I'm not considering? For the moment, I just want to collect the possibilities and their pros and cons so we can make an informed decision together.

**Comment by jbardin** [comment]
Thanks @apparentlymart, I believe those are the most reasonable options available at this time.

One note though on #1, the current behavior does _not_ fail on the first apply, because that it the case where the data source is being "created" in the state, and we defer it because of `depends_on`. The current behavior only fails when some external change causes what previously worked for that data source to return an error during refresh.

**Comment by jbardin** [comment]
This no longer happens with the new data source lifecycle in 0.14, as the data source isn't updated until we have the change information for its dependencies.
